src/main/java/StringManipulation.java

```diff
@@ -1,33 +1,133 @@
+            // corresponding character to the str
+            if(i % n != 0) {
+                str.append(stringToManipulate.charAt(i - 1));
+            }
+            // if the index is divisible by n AND maintainSpacing is true, append a space
+            else if(maintainSpacing){
+                str.append(' ');
+            }
+        }
+
+        return str.toString();
     }
 
     @Override
-    public String[] getSubStrings(int startWord, int endWord) {
-        return null;
+    public String[] getSubStrings(int startWord, int endWord){
+
+        // If the current stringToManipulate is null, throws an exception
+        if(stringToManipulate == null){
+            throw new NullPointerException("The string is null.");
+        }
+        // If the startWord or the endWord is zero or negative, OR the endWord is smaller than the startWord,
+        // throws an exception
+        if(startWord <= 0 || endWord <= 0 || startWord > endWord)
+        {
+            throw new IllegalArgumentException("Invalid argument(s) were entered.");
+        }
+
+        // Gets the word count by using split()
+        String[] words = stringToManipulate.split("[\\s.,_:?!]+");
+
+        // If the endWord is larger than the length of string, throws an exception
+        if(endWord > words.length)
+        {
+            throw new IndexOutOfBoundsException("The endWord entered is out of bounds");
+        }
+
+        // copies the words in the specified range into a new array called substring
+        String[] substring = Arrays.copyOfRange(words, startWord-1, endWord);
+
+        return substring;
     }
 
     @Override
-    public String restoreString(int[] indices) {
-        return null;
+    public String restoreString(int[] indices){
+
+        // If the current stringToManipulate is null, throws an exception
+        if(stringToManipulate == null)
+        {
+            throw new NullPointerException("The string is null");
+        }
+        // If the number of indices are not the same as the length of the string,
+        // throws an exception
+        if(indices.length != stringToManipulate.length()){
+            throw new IllegalArgumentException("The number of indices must match the length of the string.");
+        }
+
+        StringBuilder newString = new StringBuilder();
+
+        // Traverse the indices array to the end
+        for(int i = 0; i < indices.length; i++){
+            // if the index is negative, or larger than equal to the length of the string,
+            // throws an exception
+            if(indices[i] < 0 || indices[i]>= stringToManipulate.length()){
+                throw new IndexOutOfBoundsException("Each index must be between 0 and the length of the string.");
+            }
+            // otherwise append the character at the index to the new string
+            newString.append(stringToManipulate.charAt(indices[i]));
+        }
+        return newString.toString();
     }
```

Comment:
Using a StringBuilder to construct the new string is a smart move. It helps optimize performance and memory usage, especially when dealing with multiple concatenations.

Here's a small suggestion for improvement: Consider creating a separate helper method to validate the indices. By isolating the index validation logic into its own method, you can enhance code readability and maintainability. Plus, if you need to reuse this logic elsewhere in your codebase, it'll be easier to do so.

Overall, your code shows a solid understanding of exception handling and effective string manipulation. You're doing an excellent job! If you have any questions or need further assistance, feel free to ask. Your contributions are highly valued, and keep up the great work!

Reply:
Thank you for your comment with insightful information! I did not know the StringBuilder helps the optimization on the performance and memory usage, I'm glad to learn that from you:)

And thank you for the suggestion that I could create a separate helper method. I agree it'd be a good idea for reusability and maintainability. 

Based on your suggestion, I created a separate helper method to check if the string is null and if so, it throws an NullPointerException. I have the same check almost in every method so it's a good idea to separate it and reuse it. 

For other checks, I decided to leave the code as-is because I feel my current code is clear in which exceptions are handled when validating indices within a particular method and each method has slightly different conditions for validation so I cannot really reuse them.